cnxeasybake/oven.py

```diff
@@ -816,7 +816,7 @@ def eval_string_value(self, element, value):
                         u"{}".format(term.name, serialize(value)).encode(
                             'utf-8'))
 
-        if strval:
+        if strval or len(vals) == 0:
```

Comment:
I'm not totally sure about returning `['']` whenever `len(vals) == 0` because there's such a long list of conditionals above... :grimacing:

What I did instead is:

```diff
--- a/cnxeasybake/oven.py
+++ b/cnxeasybake/oven.py
@@ -719,7 +719,7 @@ class Oven():
                 pass
 
             elif type(term) is ast.StringToken:
-                strval += term.value
+                vals.append(term.value)
 
             elif type(term) is ast.IdentToken:
                 log(DEBUG, u"IdentToken as string: {}".format(
```

^ I think what we need is to make sure whenever there's a `StringToken`, the string is returned even if it's `''`.  This passes the test you added but there might be some other cases this doesn't fix...

Reply:
@karenc There are some problems associated with doing that. Directly appending to `vals` would yield incorrect orders in many cases.
For example with this markup:
```html
<html>
  <div>
  <a></a>
  </div>
</html>
```
and this ruleset:
```css
div {
  string-set: should-be-before 'am I first? ';
}

a::before {
  class: string(should-be-before) 'well yes but actually no. ';
}
```
yields before
```html
<html>
  <div>
  <a><div class="am I first? well yes but actually no. "></div></a>
  </div>
</html>
```
 and after
```html
<html>
  <div>
  <a><div class="well yes but actually no. am I first? "></div></a>
  </div>
</html>
```
which is unexpected behavior. There may be a solution in which we append to `vals` for all cases, but I'm not finding it, and it's generally causing more failures than it's fixing so far.

Another alternative is to pass the string literal as a special case like in [this branch ](https://github.com/openstax/cnx-easybake/compare/string-arg-two-alt?expand=1), but that doesn't resolve issues like #99 .

The reasoning behind this PR is that if we run `eval_string_value` on `'a'` we get `['a']`, so why should we not expect `['']` when we run it on something equivalent to `''`. One downside to this PR that I see is that now `WhitespaceToken` will now parse as `['']`. To be fair, though, cnx-easybake already did this, because it joined `vals` with `''` anyways.
For example for the same markup as above:
```css
a::before {
  class: ;
}
```
is somehow a valid ruleset before and after the change in this PR, which doesn't make sense. But I thought that since the behavior is consistent, there is not a real problem introduced by this change.

Reply:
Ok I understand why the code I pasted doesn't work because in the for loop, a lot of the statements don't use `vals.append` which causes `vals` to be out of order :(

This is really complicated...  I can't say I understand it :expressionless:

I think your fix should work because there are lots of `eval_string_value` calls that expects there to be at least one element in the list. :+1: